src/CloudNative.CloudEvents/Http/HttpListenerExtensions.cs

```diff
@@ -0,0 +1,162 @@
+
+        /// <summary>
+        /// Indicates whether this HttpListenerRequest is a web hook validation request
+        /// </summary>
+        public static bool IsWebHookValidationRequest(this HttpListenerRequest httpRequestMessage) =>
+            httpRequestMessage.HttpMethod.Equals("options", StringComparison.InvariantCultureIgnoreCase) &&
+            httpRequestMessage.Headers["WebHook-Request-Origin"] is object;
+
+        /// <summary>
+        /// Converts this listener request into a CloudEvent object, with the given extension attributes.
+        /// </summary>
+        /// <param name="httpListenerRequest">Listener request</param>
+        /// <param name="formatter"></param>
+        /// <param name="extensions">List of extension instances</param>
+        /// <returns>A CloudEvent instance or 'null' if the request message doesn't hold a CloudEvent</returns>
+        public static CloudEvent ToCloudEvent(this HttpListenerRequest httpListenerRequest,
+            ICloudEventFormatter formatter, params CloudEventAttribute[] extensionAttributes)
+        {
+            if (HasCloudEventsContentType(httpListenerRequest))
+            {
+                // FIXME: Handle no formatter being specified.
+                return formatter.DecodeStructuredEvent(httpListenerRequest.InputStream, extensionAttributes);
+            }
+            else
+            {
+                CloudEventsSpecVersion version = CloudEventsSpecVersion.Default;
+                if (httpListenerRequest.Headers[HttpUtilities.SpecVersionHttpHeader] is string versionId)
+                {
+                    version = CloudEventsSpecVersion.FromVersionId(versionId);
+                }
+
+                var cloudEvent = new CloudEvent(version, extensionAttributes);
+                var headers = httpListenerRequest.Headers;
+                foreach (var key in headers.AllKeys)
+                {
+                    string attributeName = HttpUtilities.GetAttributeNameFromHeaderName(key);
+                    if (attributeName is null || attributeName == CloudEventsSpecVersion.SpecVersionAttribute.Name)
+                    {
+                        continue;
+                    }
+                    string attributeValue = HttpUtilities.DecodeHeaderValue(headers[key]);
+                    cloudEvent.SetAttributeFromString(attributeName, attributeValue);
+                }
+
+                // TODO: Check that this doesn't come through as a header already
+                cloudEvent.DataContentType = httpListenerRequest.ContentType;
+
+                // TODO: This is a bit ugly.
+                var memoryStream = new MemoryStream();
+                httpListenerRequest.InputStream.CopyTo(memoryStream);
+                if (memoryStream.Length != 0)
+                {
+                    cloudEvent.Data = formatter.DecodeData(memoryStream.ToArray(), cloudEvent.DataContentType);
```

Comment:
If DecodeData is made to take ROM we can avoid the ToArray and instead use something like:
```cs
memoryStream.GetBuffer().AsMemory(0, (int)memoryStream.Position)
```

Reply:
Interesting. We should definitely discuss the pros and cons of taking a dependency on System.Memory. I know there are some dependencies like this that cause problems (see https://blog.marcgravell.com/2020/01/net-core-net-5-exodus-of-net-framework.html) but maybe System.Memory itself is okay.

Reply:
Sure we can discuss it - I think it is pretty widely used in newer APIs.